students/student_framllat/session04/mailroom2.py

```diff
@@ -0,0 +1,247 @@
+#!/usr/bin/env python
+"""
+mailroom assignment
+
+This version uses a dict for the main db, and exception handling to
+check input
+
+*** NOTE *** once again significant utilization of provided solution
+utilized to create functioning program. Added some modifications in
+order to grasp the concepts.
+"""
+
+import os
+import sys
+import math
+
+# handy utility to make pretty printing easier
+from textwrap import dedent
+
+
+# In memory representation of the donor database
+# using a tuple for each donor
+# -- kind of like a record in a database table
+# using a dict with a lower case version of the donor's name as the key
+# This makes it easier to have a 'normalized' key.
+#  you could get a bit fancier by having each "record" be a dict, with
+#   "name" and "donations" as keys.
+def get_donor_db():
+    return {'jed-clampett': ('Jed Clampett', [1150, 299, 2155]),
+            'jim-rockford': ('Jim Rockford', [325, 300, 125]),
+            'tony-nelson': ('Tony Nelson', [75, 100, 105]),
+            'marcia-brady': ('Marcia Brady', [15, 10]),
+            'samantha-stevens': ('Samantha Stevens', [105, 100, 200]),
+            'oscar-madison': ('Oscar Madison', [900, 450, 1200])}
+
+
+def list_donors():
+    """
+    Create a list of the donors as a string, so they can be printed
+
+    Not calling print from here makes it more flexible and easier to
+    test
+    """
+    listing = ["Donor list:"]
+    for donor in donor_db.values():
+        listing.append(donor[0])
+    return "\n".join(listing)
+    
+
+def find_donor(name):
+    """
+    Find a donor in the donor db
+
+    :param: the name of the donor
+    :returns: The donor data structure -- key is lowercase name with hyphens
+    instead of spaces - ex: Lex Luther = lex-luther
+    None if not in the donor_db
+    """
+    key = name.replace(" ", "-").strip().lower()
```

Comment:
since you need this in two places -- and the really shold be the same -- make a function that does it:

```
def name_to_key(name):
    name.replace(" ", "-").strip().lower()
```

Did I do that in my solution? I should have :-)